packages/hop-node/src/watchers/RelayWatcher.ts

```diff
@@ -0,0 +1,383 @@
+    logger.debug('l1txHash:', l1TxHash)
+
+    const isSet = await this.bridge.isTransferRootSet(transferRootHash, totalAmount)
+    if (isSet) {
+      logger.warn('checkRelayableTransferRoots already set. marking item not found.')
+      await this.db.transferRoots.update(transferRootId, { isNotFound: true })
+      return
+    }
+
+    if (this.dryMode) {
+      logger.warn(`dry: ${this.dryMode}, skipping bondTransferRoot`)
+      return
+    }
+
+    logger.debug(
+      `attempting to relay root id ${transferRootId} with destination chain ${destinationChainId} and l1TxHash ${l1TxHash}`
+    )
+
+    await this.db.transferRoots.update(transferRootId, {
+      sentRelayTxAt: Date.now()
+    })
+
+    try {
+      const tx = await this.sendTransferRootRelayTx(
+        transferRootId,
+        l1TxHash!
+      )
+      const msg = `transferRootSet dest ${destinationChainId}, tx ${tx.hash} transferRootHash: ${transferRootHash}`
+      logger.info(msg)
+      this.notifier.info(msg)
+    } catch (err) {
+      logger.error('transferRootSet error:', err.message)
+      throw err
+    }
+  }
+
+  async sendTransferRelayTx (params: any): Promise<providers.TransactionResponse> {
+    const {
+      transferId,
+      destinationChainId,
+      transferSentTxHash,
+      messageIndex
+    } = params
+    const logger = this.logger.create({ id: transferId })
+
+    logger.debug(
+      `relay transfer destinationChainId: ${destinationChainId} with messageIndex ${messageIndex}`
+    )
+    logger.debug('checkTransferSentToL2 l2Bridge.distribute')
+    return await this.sendRelayTx(transferSentTxHash, messageIndex)
+  }
+
+  async sendTransferRootRelayTx (transferRootId: string, txHash: string): Promise<providers.TransactionResponse> {
+    const logger = this.logger.create({ root: transferRootId })
+    logger.debug(
+      'relay root destinationChainId'
+    )
+    logger.debug('checkRelayableTransferRoots l2Bridge.distribute')
+    return await this.sendRelayTx(txHash)
+  }
+
+  async sendRelayTx (txHash: string, messageIndex: number = 0): Promise<providers.TransactionResponse> {
+    return await this.relayWatcher.redeemArbitrumTransaction(txHash, messageIndex)
+  }
+
+  async getMessageIndex (transferId: string, transferSentTxHash: string, transferSentTimestamp: number): Promise<number> {
+    // We need to deterministically order all the messages in an L1 tx, even if they have already been relayed
+
+    // Get all the transfers at the same time so we can get the messageIndex for each one
+    const dateFilter = {
+      fromUnix: transferSentTimestamp,
+      toUnix: transferSentTimestamp
+    }
+    const transfers: Transfer[] = await this.db.transfers.getTransfers(dateFilter)
+
+    // Get all transfers within the same L1 tx and store their log index
+    const logIndicesPerTransferId: Record<string, number> = {}
+    for (const transfer of transfers) {
+      if (transfer.transferSentTxHash !== transferSentTxHash) continue
```

Comment:
Instead of `continue` I threw here. It should be impossible for this to happen [because  `txHash` and `logIndex` are added atomically](https://github.com/hop-protocol/hop/blob/7f7cd207bba4d7dc67d247b8fc26548c42a7313b/packages/hop-node/src/watchers/SyncWatcher.ts#L417)).

If it does happen somehow, then there is a possibility that `getMessageIndex()` returns a different result for the same input, and handling that would likely require investigation. Throwing here will just make it more obvious that this needs to be looked into, IMO.

Reply:
> Instead of continue I threw here.

that makes sense however I don't see the `throw` here?

if referring to the `!` operator then that wouldn't throw during runtime

Reply:
It is [here](https://github.com/hop-protocol/hop/pull/451/files#diff-5951b7a31d1735a4930f2d972f3a02a0f4fedcfe8413d0213803edb57637b5bcR374-R376). I think there might just be some weird UI thing, but let me know if it is not showing up for you.

Reply:
I see it now 👍